internal/cri/server/stats_collector.go

```diff
@@ -0,0 +1,335 @@
+//go:build linux
+
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package server
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/containerd/cgroups/v3"
+	"github.com/containerd/cgroups/v3/cgroup1"
+	cg1 "github.com/containerd/cgroups/v3/cgroup1/stats"
+	cgroupsv2 "github.com/containerd/cgroups/v3/cgroup2"
+	cg2 "github.com/containerd/cgroups/v3/cgroup2/stats"
+	"github.com/containerd/log"
+	"github.com/containerd/typeurl/v2"
+
+	"github.com/containerd/containerd/api/services/tasks/v1"
+	criconfig "github.com/containerd/containerd/v2/internal/cri/config"
+	"github.com/containerd/containerd/v2/internal/cri/store/stats"
+	ctrdutil "github.com/containerd/containerd/v2/internal/cri/util"
+)
+
+const (
+	// defaultCollectionInterval is how often the collector fetches metrics.
+	// This matches cAdvisor's default housekeeping interval.
+	defaultCollectionInterval = 1 * time.Second
+
+	// defaultStatsAge is how long stats samples are retained.
+	defaultStatsAge = 2 * time.Minute
+)
+
+// StatsCollector periodically collects CPU stats for containers and sandboxes,
+// storing them in TimedStores. This allows UsageNanoCores to be calculated
+// from historical samples, similar to how cAdvisor works.
+type StatsCollector struct {
+	mu sync.RWMutex
+	// stores maps container/sandbox ID to their TimedStore
+	stores map[string]*stats.TimedStore
+	// interval is how often to collect stats
+	interval time.Duration
+	// statsAge is how long to keep samples
+	statsAge time.Duration
+	// maxSamples is the max number of samples per container
+	maxSamples int
+	// stopCh is used to stop the collection loop
+	stopCh chan struct{}
+	// doneCh signals when the collection loop has stopped
+	doneCh chan struct{}
+	// stopOnce ensures Stop() only closes stopCh once
+	stopOnce sync.Once
+	// service provides access to task metrics
+	service *criService
```

Comment:
`StatsCollector` is already part of `criService` struct https://github.com/containerd/containerd/pull/12629/files#diff-3bbef1ddd83b801d65af1f7d796e5193fcf66158aa7d554a22292c4e327a38a5R164. So why have the service again as a field here?